### PR TITLE
[wmcb] Allow forcing Windows node object IP 

### PR DESCRIPTION
**Pull request by sebsoto**
* Refactor to make kubelet args unit-testable 
* Allow forcing Windows node object IP

### CODE REVIEW

File: pkg/bootstrapper/bootstrapper.go [hunk 1]
```diff
@@ -522,6 +526,9 @@ func (wmcb *winNodeBootstrapper) generateInitialKubeletArgs(args map[string]stri
 	if nodeWorkerLabel, ok := args["node-labels"]; ok {
 		kubeletArgs = append(kubeletArgs, "--"+"node-labels"+"="+nodeWorkerLabel)
 	}
+	if wmcb.nodeIP != "" {
+		kubeletArgs = append(kubeletArgs, "--node-ip="+wmcb.nodeIP)
```
Comment (saifshaikh48) [inline pkg/bootstrapper/bootstrapper.go:530]: Currently it's not clear to me where `nodeIP` is being set. I assume the value will be passed in by WMCO when it runs `wmcb.exe --node-ip= ...`?

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

Left a question just to solidify my understanding. Also, I am fine with the limited scope refactor, as long as we make sure that the further work is tracked in Jira.

File: pkg/bootstrapper/bootstrapper.go (hunk 1 above)
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:530]: Yes that will passed in via WMCO when running WMCB

File: pkg/bootstrapper/bootstrapper.go [hunk 2]
```diff
@@ -522,15 +543,13 @@ func (wmcb *winNodeBootstrapper) ensureKubeletService() error {
 		Password:         "",
 		Description:      "OpenShift Kubelet",
 	}
-	// Get kubelet args
-	kubeletArgs := wmcb.getInitialKubeletArgs()
 
 	if wmcb.kubeletSVC == nil {
-		if err := wmcb.createKubeletService(c, kubeletArgs); err != nil {
+		if err := wmcb.createKubeletService(c, wmcb.kubeletArgs); err != nil {
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:555]: If you are making the actual `kubeletArgs` part of the `winNodeBootstrapper`, it should not be passed as an argument to a struct method

File: pkg/bootstrapper/bootstrapper.go [hunk 3]
```diff
@@ -336,50 +335,42 @@ func (wmcb *winNodeBootstrapper) parseIgnitionFileContents(ignitionFileContents
...
-			filesToTranslate[results[1]] = fileTranslation{
-				dest: filepath.Join(wmcb.installDir, cloudConfFilename),
-			}
-
-			// Set the --cloud-config option value
-			wmcb.kubeletArgs[cloudConfigOption] = filepath.Join(wmcb.installDir, cloudConfFilename)
+	// Check for the presence of "--cloud-config" option and if it is present append the value to
+	// filesToTranslate. This option is only present for Azure and hence we cannot assume it as a file that
+	// requires translation across clouds.
+	if cloudConfigPath, ok := args[cloudConfigOption]; ok {
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:360]: Shouldn't this now be done in `parseKubeletArgs()` and have one spot where there args are being created?

File: pkg/bootstrapper/bootstrapper.go [hunk 4]
```diff
@@ -522,6 +526,9 @@ func (wmcb *winNodeBootstrapper) generateInitialKubeletArgs(args map[string]stri
 	if nodeWorkerLabel, ok := args["node-labels"]; ok {
 		kubeletArgs = append(kubeletArgs, "--"+"node-labels"+"="+nodeWorkerLabel)
 	}
+	if wmcb.nodeIP != "" {
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:529]: Shouldn't we validate the `nodeIP` rather than blindly using it?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @sebsoto. PTAL at my comments.

File: pkg/bootstrapper/bootstrapper.go (hunk 4 above)
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:543]: Added it to the `newWinNodeBootstrapper` function

File: pkg/bootstrapper/bootstrapper.go (hunk 3 above)
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:374]: I added a TODO explaining why it is currently like that and link to the new refactor ticket https://issues.redhat.com/browse/WINC-670

Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:374]: @saifshaikh48 ticket that you asked for^

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

Thanks for working on this @sebsoto

File: pkg/bootstrapper/bootstrapper.go [hunk 5]
```diff
@@ -461,9 +485,9 @@ func (wmcb *winNodeBootstrapper) initializeKubeletFiles() error {
 	return nil
 }
 
-// getInitialKubeletArgs returns the kubelet args required during initial kubelet start up.
-// This function assumes that wmcb.kubeletArgs are populated.
-func (wmcb *winNodeBootstrapper) getInitialKubeletArgs() []string {
+// generateInitialKubeletArgs returns the kubelet args required during initial kubelet start up. args should be a map
```
Comment (mansikulkarni96) [inline pkg/bootstrapper/bootstrapper.go:500]: args -> Args

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @sebsoto 
LGTM except for a nit.

File: pkg/bootstrapper/bootstrapper.go (hunk 5 above)
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:500]: I did lowercase since I was referring to the variable `args` which is lowercase

**Review by aravindhp — APPROVED** [review]
/approve

**Comment by sebsoto** [comment]
/hold cancel
Seeing no change in flags besides --node-ip flag

**Pull request merged**